app/domain/authentication/authn_k8s/application_identity.rb

```diff
@@ -0,0 +1,181 @@
+    # the validation in ValidateApplicationIdentity is that here we validate that
+    # the application identity is configured correctly, and thus is a valid application
+    # identity. In ValidateApplicationIdentity we validate that the defined application
+    # identity is actually the correct one in kubernetes.
+    # For example, a application identity `some-namepsace/blah/some-value` is not
+    # valid and will fail the validation here. However, a application identity
+    # `some-namespace/service-account/some-value` is a valid application identity
+    # and will pass the validation here. If the host is actually running from
+    # a pod with service account `some-other-value` then it will fail the
+    # validation of ValidateApplicationIdentity
+    class ApplicationIdentity
+
+      def initialize(host_id:, host_annotations:, service_id:)
+        @host_id          = host_id
+        @host_annotations = host_annotations
+        @service_id       = service_id
+
+        validate
+      end
+
+      def namespace
+        @namespace ||= application_identity_in_annotations? ? constraint_from_annotation("namespace") : host_id_suffix[-3]
+      end
+
+      def constraints
+        @constraints ||= {
+          pod:               constraint_value("pod"),
+          service_account:   constraint_value("service_account"),
+          deployment:        constraint_value("deployment"),
+          deployment_config: constraint_value("deployment_config"),
+          stateful_set:      constraint_value("stateful_set")
+        }
+      end
+
+      def container_name
+        annotation_name = "authentication-container-name"
+        annotation_value("authn-k8s/#{@service_id}/#{annotation_name}") ||
+          annotation_value("authn-k8s/#{annotation_name}") ||
+          annotation_value("kubernetes/#{annotation_name}") ||
+          "authenticator"
+      end
+
+      def namespace_scoped?
+        constraints.compact.empty?
+      end
+
+      private
+
+      # Validates that the application identity is defined correctly
+      def validate
+        validate_permitted_scope
+        validate_namespace_constraint_exists
+        validate_constraint_combinations
+      end
+
+      # If the application identity is defined in:
+      #   - annotations: validates that all the constraints are
+      #                  valid (e.g there is no "authn-k8s/blah" annotation)
+      #   - host id: validates that the host-id has 3 parts and that the given
+      #              constraint is valid (e.g the host id is not
+      #              "namesapce/blah/some-value")
+      def validate_permitted_scope
+        application_identity_in_annotations? ? validate_permitted_annotations : validate_host_id
+      end
+
+      # This method assumes that 'validate_permitted_scope' has already run.
+      # It uses the value of 'host_id_suffix' after it is validated.
+      def validate_namespace_constraint_exists
```

Comment:
Thanks for adding this comment to point out a critical dependency in flow.  Could we move the body of this function up into to `validate_permitted_scope` method to reduce the chance that a future developer will make a change which breaks the dependency?

Reply:
good idea, thanks.